pkg/naming/metrics_query_test.go

```diff
@@ -0,0 +1,354 @@
+			extraGroupBy: []string{"extra", "groups"},
+
+			check: checks(
+				hasError(nil),
+				hasSelector("[resource extra groups]"),
+			),
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			selector, err := tc.mq.Build(tc.series, tc.resource, tc.namespace, tc.extraGroupBy, tc.names...)
+
+			if err := tc.check(selector, err); err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestBuildExternalSelector(t *testing.T) {
+	mustNewQuery := func(queryTemplate string, namespaced bool) MetricsQuery {
+		mq, err := NewMetricsQuery(queryTemplate, &resourceConverterMock{namespaced})
+		if err != nil {
+			t.Fatal(err)
+		}
+		return mq
+	}
+
+	singleReq, _ := labels.NewRequirement("foo", selection.Equals, []string{"bar"})
+	singleReqMultVals, _ := labels.NewRequirement("foo", selection.In, []string{"bar", "baz"})
+
+	tests := []struct {
+		name string
+		mq   MetricsQuery
+
+		series         string
+		namespace      string
+		groupBy        string
+		groupBySlice   []string
+		metricSelector labels.Selector
+
+		check checkFunc
+	}{
+		{
+			name: "series",
+
+			mq:             mustNewQuery(`series <<.Series>>`, false),
+			series:         "foo",
+			metricSelector: labels.NewSelector(),
+
+			check: checks(
+				hasError(nil),
+				hasSelector("series foo"),
+			),
+		},
+		{
+			name: "single GroupBy value",
+
+			mq:             mustNewQuery(`<<.GroupBy>>`, false),
+			groupBy:        "foo",
+			metricSelector: labels.NewSelector(),
+
+			check: checks(
+				hasError(nil),
+				hasSelector("foo"),
+			),
+		},
+		{
+			name: "multiple GroupBySlice values",
+
+			mq:             mustNewQuery(`<<.GroupBySlice>>`, false),
+			groupBySlice:   []string{"foo", "bar"},
+			metricSelector: labels.NewSelector(),
+
+			check: checks(
+				hasError(nil),
+				hasSelector("[foo bar]"),
+			),
+		},
+		{
+			name: "single LabelMatchers value",
+
+			mq:             mustNewQuery(`<<.LabelMatchers>>`, false),
+			metricSelector: labels.NewSelector().Add(*singleReq),
+
+			check: checks(
+				hasError(nil),
+				hasSelector(`foo="bar"`),
+			),
+		},
+		{
+			name: "multiple LabelMatchers value",
+
+			mq:             mustNewQuery(`<<.LabelMatchers>>`, false),
+			metricSelector: labels.NewSelector().Add(*singleReq, *singleReqMultVals),
+
+			check: checks(
+				hasError(nil),
+				hasSelector(`foo="bar",foo=~"bar|baz"`),
+			),
+		},
+		{
+			name: "single LabelValuesByName value",
+
+			mq:             mustNewQuery(`<<.LabelValuesByName>>`, false),
+			metricSelector: labels.NewSelector().Add(*singleReq),
+
+			check: checks(
+				hasError(nil),
+				hasSelector("map[foo:bar]"),
+			),
+		},
+		{
+			name: "multiple LabelValuesByName",
+
+			mq:             mustNewQuery(`<<.LabelValuesByName>>`, false),
+			metricSelector: labels.NewSelector().Add(*singleReq, *singleReqMultVals),
+
+			check: checks(
+				hasError(nil),
+				hasSelector("map[foo:bar|baz]"),
```

Comment:
just to avoid confusion: you are adding two metric selectors here using the same key `"foo"`. Wouldn't it be better to have two separate keys so they appear separately in the map?